manifests/plugin.pp

```diff
@@ -32,8 +32,22 @@
 ) {
   include ::jenkins::params
 
-  $plugin            = "${name}.hpi"
-  $plugin_parent_dir = inline_template('<%= @plugin_dir.split(\'/\')[0..-2].join(\'/\') %>')
+  $plugin     = "${name}.hpi"
+  $homedir    = $::jenkins::params::homedir
+  $jenkinsdir = $::jenkins::params::jenkinsdir
+  $jusername = $username ? {
+    ''      => $::jenkins::params::jenkins_user,
+    default => $username,
+  }
+  $jgroup = $group ? {
+    ''      => $::jenkins::params::jenkins_group,
+    default => $group,
+  }
+  $jplugin_dir = $plugin_dir ? {
+    ''      => $::jenkins::params::plugin_dir,
+    default => $plugin_dir,
+  }
```

Comment:
A `:pre_condition` block that pulls in `jenkins/jenkins::params`would likely be needed for those tests to work.  However, that would introduce evaluation order dependent behavior for users.  My opinion is that introducing parse order requirements at this point would be a major API break.

Unlike the `inherits` hack that works for classes, there's no way [that I'm aware of] internal to a defined type to require that a class be evaluated before it in the manifest.  It can be done external to the define by placing constraints using resource collectors (which have side effects on virtual resources) or if you know the name of the define.  I used the later approach in my selenium module.  Eg., https://github.com/jhoblitt/puppet-selenium/blob/master/manifests/hub.pp#L14-L23 Which isn't exposed to the user as its wrapped up in a class.  I think the if/else approach you've described or a custom function is the most robust approach.